File: e2e/api/src/main/kotlin/io/github/carlomicieli/trenako/catalog/db/ScaleRow.kt
/*
 *   Copyright (c) 2022-2023 (C) Carlo Micieli
 *
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 */
package io.github.carlomicieli.trenako.catalog.db

import io.github.carlomicieli.trenako.model.ScaleGauge
import io.github.carlomicieli.trenako.model.ScaleStandard
import io.github.carlomicieli.trenako.model.TrackGauge
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.math.BigDecimal
import java.time.Instant

@Table("scales")
data class ScaleRow(
    @Id
    val scaleId: String,
    val name: String,
    val ratio: BigDecimal,
    val gaugeMillimeters: BigDecimal,
    val gaugeInches: BigDecimal,
    val trackGauge: TrackGauge,
    val descriptionEn: String?,
    val descriptionIt: String?,
    val standards: List<ScaleStandard>,
    val version: Int,
    val createdAt: Instant,
    val lastModifiedAt: Instant? = null
) {
    constructor(
        scaleId: String,
        name: String,
        ratio: BigDecimal,
        gauge: ScaleGauge,
        description: Map<String, String>? = null,
        standards: List<ScaleStandard>,
        version: Int = 1,
        createdAt: Instant = Instant.now(),
        lastModifiedAt: Instant? = null
    ) : this(
        scaleId,
        name,
        ratio,
        gauge.millimeters.toBigDecimal(),
        gauge.inches.toBigDecimal(),
        gauge.trackGauge,
        description?.get("en"),
        description?.get("it"),
        standards,
        version,
        createdAt,
        lastModifiedAt
    )
}

File: crates/libs/data/src/catalog/scales/scale_row.rs
use catalog::common::TrackGauge;
use catalog::scales::scale_id::ScaleId;
use catalog::scales::standard::Standard;
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;

#[derive(Debug)]
pub struct ScaleRow {
    pub scale_id: ScaleId,
    pub name: String,
    pub ratio: Decimal,
    pub gauge_millimeters: Option<Decimal>,
    pub gauge_inches: Option<Decimal>,
    pub track_gauge: TrackGauge,
    pub description_de: Option<String>,
    pub description_en: Option<String>,
    pub description_fr: Option<String>,
    pub description_it: Option<String>,
    pub standards: Vec<Standard>,
    pub version: i32,
    pub created_at: DateTime<Utc>,
    pub last_modified_at: Option<DateTime<Utc>>,
}

#[cfg(test)]
pub mod test {
    use super::*;

    #[allow(dead_code)]
    pub fn new_scale_row(name: &str, ratio: Decimal, track_gauge: TrackGauge, created_at: DateTime<Utc>) -> ScaleRow {
        ScaleRow {
            scale_id: ScaleId::new(name),
            name: String::from(name),
            ratio,
            gauge_millimeters: None,
            gauge_inches: None,
            track_gauge,
            description_de: None,
            description_en: None,
            description_fr: None,
            description_it: None,
            standards: Vec::new(),
            version: 1,
            created_at,
            last_modified_at: None,
        }
    }
}
